rfcs/20190910-struct-tensor.md

```diff
@@ -0,0 +1,885 @@
+`StructTensor`s (for hierarchical row indices) or nested `StructTensor`s (for
+hierarchical column indices).
+
+
+## Detailed Design
+
+### `class StructTensor(CompositeTensor)`
+
+The `StructTensor` base class defines the following properties and methods:
+
+#### Accessors
+
+```python
+@property
+StructTensor.shape
+    """The static shape of this StructTensor.
+
+    The returned `TensorShape` is guaranteed to have a known rank, but the
+    individual dimension sizes may be unknown.
+    """
+```
+
+```python
+@property
+StructTensor.rank
+    """The rank of this StructTensor (`ndims`).  Guaranteed not to be `None`."""
+```
+
+```python
+StructTensor.field_names(self):
+    """Returns the string field names for this `StructTensor`.
+
+    Returns:
+      tuple of string.
+    """
+```
+
+```python
+StructTensor.field_value(self, field_name):
+    """Returns the tensor value for the specified field.
+
+    If this `StructTensor` has shape `[D1...DN]`, then the returned tensor
+    will have shape `[D1...DN, V1...VM]`, where the slice `result[d1...dN]`
+    contains the value for the scalar `StructTensor` `self[d1...dN]`.
+
+    If this is a scalar `StructTensor` (`N=0`), then the returned value is
+    field value for this structure.
+
+    Returns:
+      `Tensor`, `StructTensor`, or `RaggedTensor`.
+    """
+```
+
+#### Factories
+
+```python
+StructTensor.with_updates(self, **updates):
+    """Returns a copy of this StructTensor with one or more fields modified.
+
+    Args:
+     **updates: A mapping from string field names to new field values.  Fields
+       may be added or modified.  If `self.rank > 0`, then each new field
+       value's shape must be prefixed with `self.shape`.  E.g., if `self` is a
+       vector of structures with `n` elements, then the update values must have
+       an outer dimension of size `n`, and the `i`th element will become the
+       field value for the `i`th struct.
+
+    Returns:
+      A `StructTensor`.
+    """
+```
+
+```python
+StructTensor.without(self, *field_names):
+    """Returns a copy of this StructTensor with one or more fields removed.
+
+    Args:
+      *field_names: The names of the fields to remove.
+
+    Raises:
+      KeyError: If a specified field is not present in this StructTensor.
+    """
+```
+
+```python
+StructTensor.with_only(self, *field_names):
+    """Returns a copy of this StructTensor with only specified fields retained.
+
+    Args:
+      *field_names: The names of the fields to keep.
+
+    Raises:
+      KeyError: If a specified field is not present in this StructTensor.
+    """
+```
+
+#### Conversion
+
+```python
+StructTensor.to_py(self):
+    """Returns this StructTensor as a nested Python dict or list of dicts.
+
+    Requires that `self` was constructed in eager execution mode.  (In graph mode,
+    evaluate the `StructTensor` first, and then use `StructTensorValue.to_py()`.)
+
+    Returns:
+      A nested structure of `dict` and `list`.
+    """
+```
```

Comment:
Will there be support for conversion to other formats, e.g. `StructTensor.to_arrow()`?

Reply:
Yes; but we haven't decided whether they will be methods on StructuredTensor or free functions.